docs/sharding.md

```diff
@@ -0,0 +1,107 @@
+        
+    ```json
+    {
+        "Azure:SignalR:Endpoints:EastUs":"<ConnectionString>",
+        
+        "Azure:SignalR:Endpoints:EastUs2:Secondary":"<ConnectionString>",
+        
+        "Azure:SignalR:Endpoints:WestUs:Primary":"<ConnectionString>"
+    }
+    ```
+
+    > When you configure Azure SignalR endpoints in the App Service on Azure portal, don't forget to replace `":"` with `"__"`, the double underscore in the keys. For reasons, see [Environment variables](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-5.0#environment-variables).
+
+## Route algorithm
+
+### Default behavior
+By default, the SDK uses the [DefaultEndpointRouter](https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR/EndpointRouters/DefaultEndpointRouter.cs) to pick up endpoints.
+
+* Client routing: Randomly select one endpoint from **primary online** endpoints. If all the primary endpoints are offline, then randomly select one **secondary online** endpoint. If the selection fails again, then exception is thrown.
+
+* Server message routing: All service endpoints are returned.
+
+### Customization
+We support cutomization of route algoritm in C# language. 
+
+Here are the steps:
+* Implement a cutomized router. You can leverage information provided from [`ServiceEndpoint`](https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs) to make routing decision. See guide here: [customize-route-algorithm](https://github.com/Azure/azure-signalr/blob/dev/docs/sharding.md#customize-route-algorithm).
+
+* Register the router to DI container.
+```cs
+using Microsoft.Azure.Functions.Extensions.DependencyInjection;
+using Microsoft.Azure.SignalR;
+using Microsoft.Extensions.DependencyInjection;
+
+[assembly: FunctionsStartup(typeof(SimpleChatV3.Startup))]
+namespace SimpleChatV3
+{
+    public class Startup : FunctionsStartup
+    {
+        public override void Configure(IFunctionsHostBuilder builder)
+        {
+            builder.Services.AddSingleton<IEndpointRouter, CustomizedRouter>();
+        }
+    }
+}
+```
+
+For other languages such as Javascript, we will support route algorithm customization in the future.
+
+### Implicitly pass HttpContext to endpoint router
+If your endpoint router needs `HttpContext` to make decision, you should use an [Http trigger](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=csharp) along with `SignalRConnectionInfo` input binding, so that the `HttpContext` inside the `HttpRequest` will be passed to the router.
+
+For examle:
+```cs
+[FunctionName("negotiate")]
+public static SignalRConnectionInfo GetSignalRInfo(
+    [HttpTrigger(AuthorizationLevel.Anonymous)] HttpRequest req,
+    [SignalRConnectionInfo(HubName = "simplechat", UserId = "{query.userid}")] SignalRConnectionInfo connectionInfo)
+{
+    return connectionInfo;
```

Comment:
example of make decision?

Reply:
So add example of using `HttpContext` in router. Or combine the example of route with negotiate. This is normal negotiate function and put it in the end doesn't help understand the relationship you're mentioned. 

---
In reply to: [596887424](https://github.com/Azure/azure-functions-signalrservice-extension/pull/179#discussion_r596887424) [](ancestors = 596887424)